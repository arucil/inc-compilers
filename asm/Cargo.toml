[package]
name = "asm"
version = "0.1.0"
edition = "2018"

# See more keys and their definitions at https://doc.rust-lang.org/cargo/reference/manifest.html

[dependencies]
indexmap = "1.8.1"
enum-iterator = "0.8.1"
num-derive = "0.3.3"
num-traits = "0.2.15"
support = { path = "../support" }
