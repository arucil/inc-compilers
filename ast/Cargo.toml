[package]
name = "ast"
version = "0.1.0"
edition = "2018"

[dependencies]
pretty = "0.10.0"
support = { path = "../support" }

[dev-dependencies]
pretty_assertions = "0.7.2"
