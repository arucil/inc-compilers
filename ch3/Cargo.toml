[package]
name = "ch3"
version = "0.1.0"
edition = "2018"

# See more keys and their definitions at https://doc.rust-lang.org/cargo/reference/manifest.html

[dependencies]
ast = { path = "../ast" }
asm = { path = "../asm" }
support = { path = "../support" }
ch2 = { path = "../ch2" }
indexmap = "1.6.1"
smallvec = "1.6.1"
num-traits = "0.2.14"
maplit = "1.0.2"

[dev-dependencies]
insta = "1.7.1"
pretty_assertions = "0.7.2"
assert_cmd = "1.0.7"
predicates = "2.0.0"
scopeguard = "1.1.0"