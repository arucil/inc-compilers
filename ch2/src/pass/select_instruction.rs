use super::explicate_control::CInfo;
use asm::{Arg, Block, Instr, Label, Program, Reg};
use ast::IdxVar;
use control::*;
use indexmap::IndexSet;
use std::fmt::{self, Debug, Formatter};

pub struct Info {
  pub locals: IndexSet<IdxVar>,
}

impl Debug for Info {
  fn fmt(&self, f: &mut Formatter) -> fmt::Result {
    writeln!(f, "locals: {:?}\n", self.locals)
  }
}

pub fn select_instruction(prog: CProgram<CInfo>) -> Program<Info, IdxVar> {
  Program {
    info: Info {
      locals: prog.info.locals,
    },
    constants: Default::default(),
    blocks: prog
      .body
      .into_iter()
      .map(|(label, tail)| (label, tail_block(tail)))
      .collect(),
  }
}

fn tail_block(tail: CTail) -> Block<IdxVar> {
  let mut code = vec![];
  tail_instructions(tail, &mut code);
  Block {
    global: false,
    code,
  }
}

fn tail_instructions(mut tail: CTail, code: &mut Vec<Instr<IdxVar>>) {
  loop {
    match tail {
      CTail::Return(exp) => {
        exp_instructions(Arg::Reg(Reg::Rax), exp, code);
        code.push(Instr::Jmp(Label::Conclusion));
        return;
      }
      CTail::Seq(stmt, new_tail) => {
        stmt_instructions(stmt, code);
        tail = *new_tail;
      }
      _ => unimplemented!(),
    }
  }
}

fn stmt_instructions(stmt: CStmt, code: &mut Vec<Instr<IdxVar>>) {
  match stmt {
    CStmt::Assign { var, exp } => exp_instructions(Arg::Var(var), exp, code),
    _ => unimplemented!(),
  }
}

fn exp_instructions(
  target: Arg<IdxVar>,
  exp: CExp,
  code: &mut Vec<Instr<IdxVar>>,
) {
  match exp {
    CExp::Atom(atom) => {
      atom_instructions(target, atom, code);
    }
    CExp::Prim(CPrim::Read) => {
      code.push(Instr::Call {
        label: "read_int".to_owned(),
        arity: 0,
      });
      code.push(Instr::Mov {
        src: Arg::Reg(Reg::Rax),
        dest: target,
      });
    }
    CExp::Prim(CPrim::Neg(atom)) => {
      atom_instructions(target.clone(), atom, code);
      code.push(Instr::Neg(target));
    }
    CExp::Prim(CPrim::Add(atom1, atom2)) => {
      atom_instructions(target.clone(), atom1, code);
      let arg = match atom2 {
        CAtom::Int(n) => Arg::Imm(n),
        CAtom::Var(var) => Arg::Var(var),
        _ => unimplemented!(),
      };
      code.push(Instr::Add {
        src: arg,
        dest: target,
      });
    }
    _ => unimplemented!(),
  }
}

fn atom_instructions(
  target: Arg<IdxVar>,
  atom: CAtom,
  code: &mut Vec<Instr<IdxVar>>,
) {
  match atom {
    CAtom::Int(n) => {
      code.push(Instr::Mov {
        src: Arg::Imm(n),
        dest: target,
      });
    }
    CAtom::Var(var) => {
      code.push(Instr::Mov {
        src: Arg::Var(var),
        dest: target,
      });
    }
    _ => unimplemented!(),
  }
}

#[cfg(test)]
mod tests {
  use super::*;
  use ast::*;
  use insta::assert_snapshot;

  #[test]
  fn nested_prims() {
    let prog =
      parse(r#"(let ([x (read)] [y (+ 2 3)]) (+ (- (read)) (+ y (- 2))))"#)
        .unwrap();
    let prog = super::super::uniquify::uniquify(prog).unwrap();
    let prog =
      super::super::remove_complex_operands::remove_complex_operands(prog);
    let prog = super::super::explicate_control::explicate_control(prog);
    let result = select_instruction(prog);
    assert_snapshot!(result.to_string_pretty());
  }
}
