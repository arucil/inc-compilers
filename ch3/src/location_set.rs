use asm::{Arg, Reg};
use ast::IdxVar;
use indexmap::map::Entry;
use indexmap::IndexMap;
use num_traits::{FromPrimitive, ToPrimitive};
use smallvec::SmallVec;
use std::fmt::{self, Debug, Write};
use std::ops::Deref;

#[derive(Debug, Clone)]
pub struct LocationSet(SmallVec<[u32; 2]>);

#[derive(Debug, Clone, Copy, PartialEq, Eq, Hash)]
pub struct Location(usize);

#[derive(Debug, Clone)]
pub struct VarStore(IndexMap<IdxVar, Var>);

#[derive(Debug, Clone, Copy, PartialEq, Eq, Hash)]
pub struct Var(usize);

impl LocationSet {
  pub fn new() -> Self {
    Self(SmallVec::new())
  }

  pub fn add_reg(&mut self, reg: Reg) {
    let i = reg_index(reg);
    if self.0.is_empty() {
      self.0.push(0);
    }
    self.0[i / 32] |= 1 << (i % 32);
  }

  pub fn remove_reg(&mut self, reg: Reg) {
    let i = reg_index(reg);
    if !self.0.is_empty() {
      self.0[i / 32] &= !(1 << (i % 32));
    }
  }

  pub fn add_var(&mut self, var: Var) {
    let i = var_index(var);
    let offset = i / 32;
    if offset >= self.0.len() {
      self.0.resize(offset + 1, 0);
    }
    self.0[offset] |= 1 << (i % 32);
  }

  pub fn remove_var(&mut self, var: Var) {
    let i = var_index(var);
    let offset = i / 32;
    if offset < self.0.len() {
      self.0[offset] &= !(1 << (i % 32));
    }
  }

  pub fn remove_caller_saved_regs(&mut self) {
    self.remove_reg(Reg::Rax);
    self.remove_reg(Reg::Rcx);
    self.remove_reg(Reg::Rdx);
    self.remove_reg(Reg::Rsi);
    self.remove_reg(Reg::Rdi);
    self.remove_reg(Reg::R8);
    self.remove_reg(Reg::R9);
    self.remove_reg(Reg::R10);
    self.remove_reg(Reg::R11);
  }

  pub fn add_argument_regs(&mut self, arity: usize) {
    const ARGUMENT_REGS: [Reg; 6] =
      [Reg::Rdi, Reg::Rsi, Reg::Rdx, Reg::Rcx, Reg::R8, Reg::R9];

    assert!(arity <= 6);
    for reg in &ARGUMENT_REGS[..arity] {
      self.add_reg(*reg);
    }
  }

  pub fn iter(&self) -> LocationIter {
    let current = if self.0.len() > 0 { self.0[0] } else { 0 };
    LocationIter {
      set: self,
      index: 0,
      current,
    }
  }

  pub fn write(&self, f: &mut impl Write, var_store: &VarStore) -> fmt::Result {
    write!(f, "{{")?;
    let mut comma = false;
    for elem in self {
      if comma {
        write!(f, ", ")?;
      }
      comma = true;
      write!(f, "{:?}", elem.to_arg(var_store))?;
    }
    write!(f, "}}")
  }
}

pub struct LocationIter<'a> {
  set: &'a LocationSet,
  index: usize,
  current: u32,
}

impl<'a> Iterator for LocationIter<'a> {
  type Item = Location;

  fn next(&mut self) -> Option<Location> {
    while self.current == 0 {
      self.index += 1;
      if self.index >= self.set.0.len() {
        return None;
      }
      self.current = self.set.0[self.index];
    }
    let i = self.current.trailing_zeros() as usize + self.index * 32;
    self.current &= self.current - 1;
    Some(Location(i))
  }
}

impl<'a> IntoIterator for &'a LocationSet {
  type Item = Location;
  type IntoIter = LocationIter<'a>;

  fn into_iter(self) -> Self::IntoIter {
    self.iter()
  }
}

fn reg_index(reg: Reg) -> usize {
  reg.to_usize().unwrap()
}

fn var_index(var: Var) -> usize {
  var.0 + 16
}

impl VarStore {
  pub fn new() -> Self {
    Self(IndexMap::new())
  }

  pub fn get(&mut self, var: IdxVar) -> Var {
    let len = self.0.len();
    match self.0.entry(var) {
      Entry::Occupied(var) => *var.get(),
      Entry::Vacant(entry) => *entry.insert(Var(len)),
    }
  }
}

impl Deref for VarStore {
  type Target = IndexMap<IdxVar, Var>;

  fn deref(&self) -> &Self::Target {
    &self.0
  }
}

impl Var {
  pub fn to_arg(&self, var_store: &VarStore) -> Arg<IdxVar> {
    Arg::Var(var_store.get_index(self.0 - 16).unwrap().0.clone())
  }
}

impl Location {
  pub fn to_arg(&self, var_store: &VarStore) -> Arg<IdxVar> {
    if self.0 < 16 {
      Arg::Reg(Reg::from_usize(self.0).unwrap())
    } else {
      Arg::Var(var_store.0.get_index(self.0 - 16).unwrap().0.clone())
    }
  }

  pub fn from_arg(arg: Arg<IdxVar>, var_store: &mut VarStore) -> Option<Self> {
    match arg {
      Arg::Deref(reg, _) | Arg::Reg(reg) => Some(reg.into()),
      Arg::Var(var) => Some(Self(var_store.get(var).0 + 16)),
      Arg::Imm(_) => None,
      Arg::ByteReg(_) => unimplemented!(),
    }
  }

  pub fn to_reg(&self) -> Option<Reg> {
    Reg::from_usize(self.0)
  }

  pub fn to_var(&self) -> Option<Var> {
    if self.0 >= 16 {
      Some(Var(self.0 - 16))
    } else {
      None
    }
  }
}

impl From<Reg> for Location {
  fn from(reg: Reg) -> Self {
    Self(reg.to_usize().unwrap())
  }
}
