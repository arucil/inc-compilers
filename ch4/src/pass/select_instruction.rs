use super::explicate_control::CInfo;
use asm::{Arg, Block, ByteReg, Instr, Label, Program, Reg};
use ast::IdxVar;
use ch2::pass::select_instruction::Info;
use control::*;
use indexmap::IndexMap;

pub fn select_instruction(prog: CProgram<CInfo>) -> Program<Info, IdxVar> {
  let mut constants = IndexMap::default();
  Program {
    info: Info {
      locals: prog.info.locals,
    },
    blocks: prog
      .body
      .into_iter()
      .map(|(label, tail)| (label, tail_block(tail, &mut constants)))
      .collect(),
    constants,
  }
}

fn tail_block(
  tail: CTail,
  constants: &mut IndexMap<String, String>,
) -> Block<IdxVar> {
  let mut code = vec![];
  tail_instructions(tail, &mut code, constants);
  Block {
    global: false,
    code,
  }
}

fn tail_instructions(
  mut tail: CTail,
  code: &mut Vec<Instr<IdxVar>>,
  constants: &mut IndexMap<String, String>,
) {
  loop {
    match tail {
      CTail::Return(exp) => {
        exp_instructions(Arg::Reg(Reg::Rax), exp, code);
        code.push(Instr::Jmp(Label::Conclusion));
        return;
      }
      CTail::Seq(stmt, new_tail) => {
        stmt_instructions(stmt, code, constants);
        tail = *new_tail;
      }
      CTail::Goto(label) => {
        code.push(Instr::Jmp(label));
        return;
      }
      CTail::If(cmp, arg1, arg2, conseq, alt) => {
        let arg1 = atom_to_arg(arg1);
        let arg2 = atom_to_arg(arg2);
        code.push(Instr::Cmp {
          dest: arg1,
          src: arg2,
        });
        code.push(Instr::JumpIf(cmp.into(), conseq));
        code.push(Instr::Jmp(alt));
        return;
      }
      _ => unimplemented!(),
    }
  }
}

fn stmt_instructions(
  stmt: CStmt,
  code: &mut Vec<Instr<IdxVar>>,
  constants: &mut IndexMap<String, String>,
) {
  match stmt {
    CStmt::Assign { var, exp } => exp_instructions(Arg::Var(var), exp, code),
    CStmt::Print {
      val,
      ty: CType::Int,
    } => {
      exp_instructions(Arg::Reg(Reg::Rax), val, code);
      code.push(Instr::Call("print_int".to_owned(), 0));
    }
    CStmt::Print {
      val,
      ty: CType::Bool,
    } => {
      exp_instructions(Arg::Reg(Reg::Rax), val, code);
      code.push(Instr::Call("print_bool".to_owned(), 0));
    }
    CStmt::Print {
      val,
      ty: CType::Str,
    } => match val {
      CExp::Atom(CAtom::Str(s)) => {
        let label = format!("const_{}", constants.len() + 1);
        let len = s.len();
        constants.insert(label.clone(), s);
        code.push(Instr::Mov {
          src: Arg::Label(label),
          dest: Arg::Reg(Reg::Rsi),
        });
        code.push(Instr::Mov {
          src: Arg::Imm(len as i64),
          dest: Arg::Reg(Reg::Rdx),
        });
        code.push(Instr::Call("print_str".to_owned(), 0));
      }
      _ => unreachable!(),
    },
    CStmt::NewLine => code.push(Instr::Call("print_newline".to_owned(), 0)),
    CStmt::Read => code.push(Instr::Call("read_int".to_owned(), 0)),
    _ => unimplemented!("{:?}", stmt),
  }
}

fn exp_instructions(
  target: Arg<IdxVar>,
  exp: CExp,
  code: &mut Vec<Instr<IdxVar>>,
) {
  match exp {
    CExp::Atom(atom) => atom_instructions(target, atom, code),
    CExp::Prim(prim) => prim_instructions(target, prim, code),
  }
}

fn prim_instructions(
  target: Arg<IdxVar>,
  prim: CPrim,
  code: &mut Vec<Instr<IdxVar>>,
) {
  match prim {
    CPrim::Read => {
      code.push(Instr::Call("read_int".to_owned(), 0));
      code.push(Instr::Mov {
        src: Arg::Reg(Reg::Rax),
        dest: target,
      });
    }
    CPrim::Neg(atom) => {
      atom_instructions(target.clone(), atom, code);
      code.push(Instr::Neg(target));
    }
    CPrim::Add(atom1, atom2) => {
      atom_instructions(target.clone(), atom1, code);
      let arg = atom_to_arg(atom2);
      code.push(Instr::Add {
        src: arg,
        dest: target,
      });
    }
    CPrim::Sub(atom1, atom2) => {
      atom_instructions(target.clone(), atom1, code);
      let arg = atom_to_arg(atom2);
      code.push(Instr::Sub {
        src: arg,
        dest: target,
      });
    }
    CPrim::Not(atom) => {
      atom_instructions(target.clone(), atom, code);
      code.push(Instr::Xor {
        src: Arg::Imm(1),
        dest: target,
      });
    }
    CPrim::Cmp(cmp, atom1, atom2) => {
      let arg1 = atom_to_arg(atom1);
      let arg2 = atom_to_arg(atom2);
      code.push(Instr::Cmp {
        dest: arg1,
        src: arg2,
      });
      code.push(Instr::SetIf(cmp.into(), Arg::ByteReg(ByteReg::Al)));
      code.push(Instr::Movzx {
        src: Arg::ByteReg(ByteReg::Al),
        dest: target,
      });
    }
    _ => unimplemented!(),
  }
}

fn atom_to_arg(atom: CAtom) -> Arg<IdxVar> {
  match atom {
    CAtom::Int(n) => Arg::Imm(n),
    CAtom::Var(var) => Arg::Var(var),
    CAtom::Bool(true) => Arg::Imm(1),
    CAtom::Bool(false) => Arg::Imm(0),
    _ => unimplemented!(),
  }
}

fn atom_instructions(
  target: Arg<IdxVar>,
  atom: CAtom,
  code: &mut Vec<Instr<IdxVar>>,
) {
  match atom {
    CAtom::Int(n) => {
      code.push(Instr::Mov {
        src: Arg::Imm(n),
        dest: target,
      });
    }
    CAtom::Var(var) => {
      code.push(Instr::Mov {
        src: Arg::Var(var),
        dest: target,
      });
    }
    CAtom::Bool(true) => {
      code.push(Instr::Mov {
        src: Arg::Imm(1),
        dest: target,
      });
    }
    CAtom::Bool(false) => {
      code.push(Instr::Mov {
        src: Arg::Imm(0),
        dest: target,
      });
    }
    _ => unimplemented!(),
  }
}

#[cfg(test)]
mod tests {
  use super::*;
  use ast::*;
  use insta::assert_snapshot;

  #[test]
  fn nested_prims() {
    let prog =
      parse(r#"(let ([x (read)] [y (+ 2 3)]) (+ (- (read)) (+ y (- 2))))"#)
        .unwrap();
    let prog = super::super::uniquify::uniquify(prog);
    let prog = super::super::anf::anf(prog);
    let prog = super::super::explicate_control::explicate_control(prog);
    let result = select_instruction(prog);
    assert_snapshot!(result.to_string_pretty());
  }

  #[test]
  fn nested_if() {
    let prog = ast::parse(
      r#"
(let ([x (read)]
      [y (read)])
  (if (if (< x 1)
        (eq? x 0)
        (eq? x 2))
    (+ y 2)
    (+ y 10)))
    "#,
    )
    .unwrap();
    let prog = super::super::uniquify::uniquify(prog);
    let prog = super::super::anf::anf(prog);
    let prog = super::super::explicate_control::explicate_control(prog);
    let result = select_instruction(prog);
    assert_snapshot!(result.to_string_pretty());
  }

  #[test]
  fn if_in_init() {
    let prog = ast::parse(
      r#"(let ([x (if (>= (read) 3) 10 77)]) (if (not (eq? x 10)) 41 2))"#,
    )
    .unwrap();
    let prog = super::super::uniquify::uniquify(prog);
    let prog = super::super::anf::anf(prog);
    let prog = super::super::explicate_control::explicate_control(prog);
    let result = select_instruction(prog);
    assert_snapshot!(result.to_string_pretty());
  }

  #[test]
  fn complex_if() {
    let prog = ast::parse(
      r#"
(let ([x (read)])
  (if (> x 100)
    (let ([y (read)])
      (if (or (and (>= x y) (< (- x y) 10))
              (and (< x y) (< (- y x) 10)))
        -1
        (- y)))
    (if (and (> x 40) (< x 60))
      5000
      x)))
      "#,
    )
    .unwrap();
    let prog = super::super::shrink::shrink(prog);
    let prog = super::super::uniquify::uniquify(prog);
    let prog = super::super::anf::anf(prog);
    let prog = super::super::explicate_control::explicate_control(prog);
    let result = select_instruction(prog);
    assert_snapshot!(result.to_string_pretty());
  }
}
