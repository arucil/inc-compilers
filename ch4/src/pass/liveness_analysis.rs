use asm::{Arg, Block, Instr, Label, Program};
use ast::IdxVar;
use ch2::pass::select_instruction::Info as OldInfo;
use ch3::location_set::{LocationSet, VarStore};
use indexmap::{IndexMap, IndexSet};
use petgraph::algo::toposort;
use petgraph::graph::NodeIndex;
use petgraph::Graph;
use std::collections::HashMap;

pub struct Info {
  pub locals: IndexSet<IdxVar>,
  pub live: IndexMap<Label, Vec<LocationSet>>,
  pub var_store: VarStore,
}

pub fn analyze_liveness(
  prog: Program<OldInfo, IdxVar>,
  label_live: HashMap<Label, LocationSet>,
) -> Program<Info, IdxVar> {
  let mut state = AnalysisState {
    var_store: VarStore::new(),
    label_live,
  };

  let live = sort_blocks(&prog.blocks)
    .iter()
    .map(|(label, block)| {
      let live = state.block_liveness(block);
      state.label_live.insert(*label, live[0].clone());
      (label.clone(), live)
    })
    .collect();

  Program {
    info: Info {
      locals: prog.info.locals,
      live,
      var_store: state.var_store,
    },
    blocks: prog.blocks,
  }
}

fn sort_blocks(
  blocks: &[(Label, Block<IdxVar>)],
) -> Vec<(Label, &Block<IdxVar>)> {
  let mut graph = Graph::new();
  let mut nodes = HashMap::<Label, NodeIndex>::new();
  for (label, block) in blocks {
    let ix = graph.add_node((*label, block));
    nodes.insert(*label, ix);
  }

  for (_, node) in &nodes {
    let last = graph[*node].1.code.last().unwrap().clone();
    if let Instr::Jmp(label) = last {
      if let Some(&node1) = nodes.get(&label) {
        graph.add_edge(*node, node1, ());
      }
      let code = &graph[*node].1.code;
      if code.len() > 1 {
        if let Instr::JumpIf(_, label) = code[code.len() - 2].clone() {
          graph.add_edge(*node, nodes[&label], ());
        }
      }
    }
  }

  toposort(&graph, None)
    .unwrap()
    .into_iter()
    .rev()
    .map(|node| graph[node])
    .collect()
}

struct AnalysisState {
  var_store: VarStore,
  label_live: HashMap<Label, LocationSet>,
}

impl AnalysisState {
  fn block_liveness(&mut self, block: &asm::Block<IdxVar>) -> Vec<LocationSet> {
    let mut set = vec![LocationSet::new(); block.code.len() + 2];
    for (i, ins) in block.code.iter().enumerate().rev() {
      set[i] = set[i + 1].clone();
      self.instr_liveness(ins, &mut set[i]);
    }
    set.pop();
    set
  }

  fn instr_liveness(&mut self, ins: &Instr<IdxVar>, before: &mut LocationSet) {
    match ins {
      Instr::Ret => {}
      Instr::Syscall => {}
      Instr::Pop(arg) => {
        self.remove_arg(before, arg);
      }
      Instr::Push(arg) => {
        self.add_arg(before, arg);
      }
      Instr::Add { src, dest }
      | Instr::Sub { src, dest }
      | Instr::Xor { src, dest } => {
        self.add_arg(before, src);
        self.add_arg(before, dest);
      }
      Instr::Mov { src, dest } | Instr::Movzx { src, dest } => {
        self.remove_arg(before, dest);
        self.add_arg(before, src);
      }
      Instr::Neg(dest) => {
        self.add_arg(before, dest);
      }
      Instr::Cmp { src, dest } => {
        self.add_arg(before, src);
        self.add_arg(before, dest);
      }
      Instr::Jmp(label) => {
        *before = self.label_live[label].clone();
      }
      Instr::SetIf(_, dest) => {
        self.remove_arg(before, dest);
      }
      Instr::JumpIf(_, label) => {
        *before |= &self.label_live[label];
      }
      Instr::Call(_, arity) => {
        before.remove_caller_saved_regs();
        before.add_argument_regs(*arity);
      }
      _ => unimplemented!("{:?}", ins),
    }
  }

  fn remove_arg(&mut self, set: &mut LocationSet, arg: &Arg<IdxVar>) {
    match arg {
      Arg::Deref(reg, _) | Arg::Reg(reg) => {
        set.remove_reg(*reg);
      }
      Arg::ByteReg(reg) => {
        set.remove_reg((*reg).into());
      }
      Arg::Var(var) => {
        let var = self.var_store.get(var.clone());
        set.remove_var(var);
      }
      _ => {}
    }
  }

  fn add_arg(&mut self, set: &mut LocationSet, arg: &Arg<IdxVar>) {
    match arg {
      Arg::Deref(reg, _) | Arg::Reg(reg) => {
        set.add_reg(*reg);
      }
      Arg::ByteReg(reg) => {
        set.add_reg((*reg).into());
      }
      Arg::Var(var) => {
        let var = self.var_store.get(var.clone());
        set.add_var(var);
      }
      _ => {}
    }
  }
}

#[cfg(test)]
mod tests {
  use super::*;
  use asm::Label;
  use insta::assert_snapshot;
  use maplit::hashmap;

  trait ShowLiveness {
    fn show(&self) -> String;
  }

  impl ShowLiveness for Program<Info, IdxVar> {
    fn show(&self) -> String {
      let mut buf = String::new();
      for (label, block) in &self.blocks {
        buf += &format!("{:?}:\n", label);
        let live = &self.info.live[label];
        for i in 0..block.code.len() {
          buf += "                    ";
          live[i].write(&mut buf, &self.info.var_store).unwrap();
          buf += "\n";
          buf += &format!("    {:?}", block.code[i]);
          buf += "\n";
        }
        buf += "                    ";
        live
          .last()
          .unwrap()
          .write(&mut buf, &self.info.var_store)
          .unwrap();
        buf += "\n\n";
      }
      buf
    }
  }

  #[test]
  fn example_in_book() {
    use asm::Reg::*;
    let blocks = asm::parse_blocks(
      |s| IdxVar::new(s),
      r#"
      start:
        mov v, 1
        mov w, 42
        mov x, v
        add x, 7
        mov y, x
        mov z, x
        add z, w
        mov t, y
        neg t
        mov rax, z
        add rax, t
        jmp conclusion
    "#,
    );
    let label_live = hashmap! {
      Label::Conclusion => {
        let mut set = LocationSet::new();
        set.add_reg(Rax);
        set.add_reg(Rsp);
        set
      }
    };
    let prog = Program {
      info: OldInfo {
        locals: IndexSet::new(),
      },
      blocks,
    };
    let result = analyze_liveness(prog, label_live);

    assert_snapshot!(result.show());
  }

  #[test]
  fn push_pop() {
    use asm::Reg::*;
    let blocks = asm::parse_blocks(
      |s| IdxVar::new(s),
      r#"
      start:
        push x
        mov w, rbx
        pop rbx
        add x, w
        jmp conclusion
    "#,
    );
    let label_live = hashmap! {
      Label::Conclusion => {
        let mut set = LocationSet::new();
        set.add_reg(Rax);
        set.add_reg(Rsp);
        set
      }
    };
    let prog = Program {
      info: OldInfo {
        locals: IndexSet::new(),
      },
      blocks,
    };
    let result = analyze_liveness(prog, label_live);

    assert_snapshot!(result.show());
  }

  #[test]
  fn call() {
    use asm::Reg::*;
    let blocks = asm::parse_blocks(
      |s| IdxVar::new(s),
      r#"
      start:
        pop rdi
        pop rsi
        push x
        mov w, rbx
        call foo, 3
        add w, rax
        jmp conclusion
    "#,
    );
    let label_live = hashmap! {
      Label::Conclusion => {
        let mut set = LocationSet::new();
        set.add_reg(Rax);
        set.add_reg(Rsp);
        set
      }
    };
    let prog = Program {
      info: OldInfo {
        locals: IndexSet::new(),
      },
      blocks,
    };
    let result = analyze_liveness(prog, label_live);

    assert_snapshot!(result.show());
  }

  #[test]
  fn epilogue() {
    let blocks = asm::parse_blocks(
      |s| IdxVar::new(s),
      r#"
      start:
        mov rsp, rbp
        pop rbp
        call print_int
        call print_newline
        mov rax, 60
        mov rdi, 0
        syscall
    "#,
    );
    let label_live = HashMap::new();
    let prog = Program {
      info: OldInfo {
        locals: IndexSet::new(),
      },
      blocks,
    };
    let result = analyze_liveness(prog, label_live);

    assert_snapshot!(result.show());
  }

  #[test]
  fn nested_if() {
    use asm::Reg::*;
    let blocks = asm::parse_blocks(
      |s| IdxVar::new(s),
      r#"
      start:
        call read_int
        mov x, rax
        call read_int
        mov y, rax
        cmp x, 1
        jl block2
        jmp block3
      block0:
        mov rax, y
        add rax, 2
        jmp conclusion
      block1:
        mov rax, z
        add rax, 10
        jmp conclusion
      block2:
        cmp x, 0
        je block0
        jmp block1
      block3:
        cmp x, 2
        je block0
        jmp block1
    "#,
    );
    let label_live = hashmap! {
      Label::Conclusion => {
        let mut set = LocationSet::new();
        set.add_reg(Rax);
        set.add_reg(Rsp);
        set
      }
    };
    let prog = Program {
      info: OldInfo {
        locals: IndexSet::new(),
      },
      blocks,
    };
    let result = analyze_liveness(prog, label_live);

    assert_snapshot!(result.show());
  }

  #[test]
  fn if_in_init() {
    use asm::Reg::*;
    let blocks = asm::parse_blocks(
      |s| IdxVar::new(s),
      r#"
start:
    call read_int
    mov tmp.0, rax
    cmp tmp.0, 3
    jge block3
    jmp block4
block0:
    mov rax, 2
    jmp conclusion
block1:
    mov rax, 41
    jmp conclusion
block2:
    cmp x.0, 10
    sete al
    movzx tmp.1, al
    cmp tmp.1, 0
    je block1
    jmp block0
block3:
    mov x.0, 10
    jmp block2
block4:
    mov x.0, 77
    jmp block2

    "#,
    );
    let label_live = hashmap! {
      Label::Conclusion => {
        let mut set = LocationSet::new();
        set.add_reg(Rax);
        set.add_reg(Rsp);
        set
      }
    };
    let prog = Program {
      info: OldInfo {
        locals: IndexSet::new(),
      },
      blocks,
    };
    let result = analyze_liveness(prog, label_live);

    assert_snapshot!(result.show());
  }

  #[test]
  fn complex_if() {
    use asm::Reg::*;
    let blocks = asm::parse_blocks(
      |s| IdxVar::new(s),
      r#"
start:
    call read_int
    mov x.0, rax
    cmp x.0, 100
    jg block8
    jmp block9
block0:
    mov rax, -1
    jmp conclusion
block1:
    mov rax, y.1
    neg rax
    jmp conclusion
block2:
    mov tmp.1, y.1
    sub tmp.1, x.0
    cmp tmp.1, 10
    jl block0
    jmp block1
block3:
    cmp x.0, y.1
    jl block2
    jmp block1
block4:
    mov tmp.0, x.0
    sub tmp.0, y.1
    cmp tmp.0, 10
    jl block0
    jmp block3
block5:
    mov rax, 5000
    jmp conclusion
block6:
    mov rax, x.0
    jmp conclusion
block7:
    cmp x.0, 60
    jl block5
    jmp block6
block8:
    call read_int
    mov y.1, rax
    cmp x.0, y.1
    jge block4
    jmp block3
block9:
    cmp x.0, 40
    jg block7
    jmp block6
    "#,
    );
    let label_live = hashmap! {
      Label::Conclusion => {
        let mut set = LocationSet::new();
        set.add_reg(Rax);
        set.add_reg(Rsp);
        set
      }
    };
    let prog = Program {
      info: OldInfo {
        locals: IndexSet::new(),
      },
      blocks,
    };
    let result = analyze_liveness(prog, label_live);

    assert_snapshot!(result.show());
  }
}
