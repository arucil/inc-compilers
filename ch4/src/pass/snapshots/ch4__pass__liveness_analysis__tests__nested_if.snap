---
source: ch4/src/pass/liveness_analysis.rs
expression: result.show()

---
start:
                    {rsp, z.0}
    call read_int
                    {rsp, rax, z.0}
    mov x.0, rax
                    {rsp, z.0, x.0}
    call read_int
                    {rsp, rax, z.0, x.0}
    mov y.0, rax
                    {rsp, z.0, y.0, x.0}
    cmp x.0, 1
                    {rsp, z.0, y.0, x.0}
    jl block2
                    {rsp, z.0, y.0, x.0}
    jmp block3
                    {}

block0:
                    {rsp, y.0}
    mov rax, y.0
                    {rsp, rax}
    add rax, 2
                    {rsp, rax}
    jmp conclusion
                    {}

block1:
                    {rsp, z.0}
    mov rax, z.0
                    {rsp, rax}
    add rax, 10
                    {rsp, rax}
    jmp conclusion
                    {}

block2:
                    {rsp, z.0, y.0, x.0}
    cmp x.0, 0
                    {rsp, z.0, y.0}
    je block0
                    {rsp, z.0}
    jmp block1
                    {}

block3:
                    {rsp, z.0, y.0, x.0}
    cmp x.0, 2
                    {rsp, z.0, y.0}
    je block0
                    {rsp, z.0}
    jmp block1
                    {}


